substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java

```diff
@@ -861,7 +861,9 @@ private void setupNativeImage(String imageName, OptionValues options, Map<Method
 
                 if (!(NativeImageOptions.ExitAfterRelocatableImageWrite.getValue() && CAnnotationProcessorCache.Options.UseCAPCache.getValue())) {
                     CCompilerInvoker compilerInvoker = CCompilerInvoker.create(ImageSingletons.lookup(TemporaryBuildDirectoryProvider.class).getTemporaryBuildDirectory());
-                    compilerInvoker.verifyCompiler();
+                    if (!CAnnotationProcessorCache.Options.DontRunCAPQueryCode.getValue()) {
+                        compilerInvoker.verifyCompiler();
+                    }
                     ImageSingletons.add(CCompilerInvoker.class, compilerInvoker);
```

Comment:
If you don't run the compiler driver because you exit after QueryCodeGeneration you should not even need to register the singleton. See https://github.com/oracle/graal/pull/2674/files#r464502705

Reply:
Changed